modules/gmc-agent/src/main/java/org/gridgain/action/controller/QueryActionsController.java

```diff
@@ -0,0 +1,193 @@
+
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.MANAGEMENT_POOL;
+import static org.gridgain.utils.QueryUtils.fetchSqlQueryRows;
+import static org.gridgain.utils.QueryUtils.getColumns;
+import static org.gridgain.utils.QueryUtils.prepareQuery;
+
+/**
+ * Query actions controller.
+ */
+@ActionController("QueryActions")
+public class QueryActionsController {
+    /** Context. */
+    private final GridKernalContext ctx;
+
+    /** Query registry. */
+    private final QueryHolderRegistry qryRegistry;
+
+    /** Query process. */
+    private GridQueryProcessor qryProc;
+
+    /** Logger. */
+    private IgniteLogger log;
+
+    /**
+     * @param ctx Context.
+     */
+    public QueryActionsController(GridKernalContext ctx) {
+        this.ctx = ctx;
+        this.log = ctx.log(QueryActionsController.class);
+
+        qryProc = ctx.query();
+        qryRegistry = new QueryHolderRegistry(ctx, Duration.ofMinutes(5));
+    }
+
+    /**
+     * Cancel query by query ID.
+     *
+     * @param qryId Query id.
+     */
+    public CompletableFuture<Void> cancel(String qryId) {
+        qryRegistry.cancelQuery(qryId);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * @param arg Argument.
+     * @return Next page with result.
+     */
+    public CompletableFuture<QueryResult> nextPage(NextPageQueryArgument arg) {
+        final CompletableFuture<QueryResult> fut = new CompletableFuture<>();
+
+        ctx.closure().runLocalSafe(() -> {
+            try {
+                fut.complete(fetchResult(arg.getQueryId(), arg.getCursorId(), arg.getPageSize()));
+            }
+            catch (Throwable e) {
+                fut.completeExceptionally(e);
+            }
+        }, MANAGEMENT_POOL);
+
+        return fut;
+    }
+
+    /**
+     * @param arg Argument.
+     * @return List of query results.
+     */
+    public CompletableFuture<List<QueryResult>> execute(QueryArgument arg) {
+        final CompletableFuture<List<QueryResult>> fut = new CompletableFuture<>();
+
+        ctx.closure().runLocalSafe(() -> {
+            String qryId = arg.getQueryId();
+            qryRegistry.cancelQuery(qryId);
+
+            QueryHolder qryHolder = qryRegistry.createQueryHolder(qryId);
+
+            try {
+                if (log.isDebugEnabled())
+                    log.debug("Operation started with subject: " + ctx.security().securityContext().subject());
+
+                SqlFieldsQuery qry = prepareQuery(arg);
+                GridCacheContext cctx = !F.isEmpty(arg.getCacheName())
+                        ? ctx.cache().cache(arg.getCacheName()).context()
+                        : null;
+
+                List<QueryResult> results = new ArrayList<>();
+                for (FieldsQueryCursor<List<?>> cur : qryProc.querySqlFields(cctx, qry, null, true, false, qryHolder.getCancelHook())) {
+                    CursorHolder curHolder = qryRegistry.addCursor(qryId, cur);
```

Comment:
If we will save cursors in fetchResult method we will not to reuse it in nextPage method, because cursor iterator was already fetched, and we still have to close the cursor whether we saved it or not.